### PR TITLE
Be able to use S3 asset source using IAM role authentication

### PR DESCRIPTION
**Pull request by astoltz**
Amazon EC2 machines can have an IAM role assigned which can grant access to S3 buckets. A major advantage to this is automatically rotating keys.

### CODE REVIEW

File: app/lib/S3.php [hunk 1]
```diff
@@ -215,6 +225,36 @@ public static function setAuth($accessKey, $secretKey)
...
 	}
 
+	/**
+	 * Set AWS access key and secret key from an IAM role.
+	 *
+	 * @throws Exception
+	 */
+	public static function setAuthFromIam()
+	{
+		$rest = new S3Request('GET', '', '/latest/meta-data/iam/info', '169.254.169.254');
```
Comment (bakura10) [inline app/lib/S3.php]: Why not using the official AWS PHP SDK? It takes care of this automatically ;).

**Comment by bakura10** [comment]
+1, this is very useful, I'm using IAM role all the time on AWS!

File: app/lib/S3.php (hunk 1 above)
Comment (astoltz) [inline app/lib/S3.php]: My guess is they're concerned about size. This library takes care of what Craft needs (minus IAM roles) without all the additional libraries for every other AWS product.

**Comment by takobell** [comment]
S3 has been moved to it's own separate plugin in Craft 3.  Future pull requests should be directed here: https://github.com/craftcms/aws-s3

**Pull request closed without merging**